### ISSUE TITLE
[Bug] `flags.warn_error_options.silence` should be settable `dbt_project.yml`

### ISSUE THREAD

**Issue by jtcohen6** [dbt-labs/dbt-core#10160]
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

Missing `silence` here:

https://github.com/dbt-labs/dbt-core/blob/32a7f82772d90a4b9ab443908f5f4455d0aa2c73/core/dbt/cli/flags.py#L53-L62

### Expected Behavior

```python
def convert_config(config_name, config_value):
    """Convert the values from config and original set_from_args to the correct type."""
    ret = config_value
    if config_name.lower() == "warn_error_options" and type(config_value) == dict:
        ret = WarnErrorOptions(
            include=config_value.get("include", []),
            exclude=config_value.get("exclude", []),
            valid_error_names=ALL_EVENT_NAMES,
            silence=config_value.get("silence", []),  # this line is missing above
        )
    return ret
```

### Steps To Reproduce

1. Create a project with a test in it
```yml
# dbt_project.yml
flags:
  warn_error_options:
    silence:
      - TestsConfigDeprecation

tests:  # has been renamed to data_tests
  +enabled: true
```
2. `dbt parse` -> still see the deprecation warning ("The `tests` config has been renamed to `data_tests`")
3. Apply the patch shown above -> `dbt parse` -> no deprecation warning

### Environment

```markdown
- Python: 3.10.11
- dbt: 1.8.0
```

**Comment by kokorin** [comment]
I want to add some reasoning to this issue.
We checking if we can update to DBT 1.8, and we changed our configs from `tests` to `data_tests`.

The problem is that we use several DBT packages and those packages still use `tests`.
And I believe that DBT package developers will not update from `tests` to `data_tests` because they want to support compatibility with previous DBT versions.

May be it even makes sense to allow warning suppression based on package name, rather than suppressing globally for the whole project.

**Comment by jtcohen6** [comment]
@kokorin Appreciate that rationale! We have a separate issue for scoping that deprecation warning more narrowly:
- https://github.com/dbt-labs/dbt-core/issues/9835

**Comment by ChenyuLInx** [comment]
https://github.com/dbt-labs/dbt-core/pull/10058/files#diff-9a7ad4b74c0941bce1368c1e8e12f0745bf0152b00f92df206375702b46e3a8eR133-R140 
There should be an integration test catching this. 
This might be because include is being required now.(something we potentially want to change).

**Comment by QMalcolm** [comment]
> This might be because include is being required now.(something we potentially want to change).

For clarity, specifying `include` as part of `warn_error_options` was _always_ required previously. Ever since the [initial implementation](https://github.com/dbt-labs/dbt-core/blob/d6ac340df09762040ef737c0463c418b78be5818/core/dbt/helper_types.py#L96), `include` was required when specifying `warn_error_options`. Additionally, that made sense in a world where only `include`/`exclude` (now `error`/`warn`) were the only options. Now that `silence` exists, that logical requirement kind of breaks down.

**Comment by QMalcolm** [comment]
About my previous comment, I was wrong. The _object_ requires `include` be specified, [it is not optional](https://github.com/dbt-labs/dbt-common/blob/00fbf918f55e72a9a6cf08767e07ee55e15cdc93/dbt_common/helper_types.py#L69). However, in the instantiation in core, we always [default it to a list](https://github.com/dbt-labs/dbt-core/blob/ca163c3d6ecaf989e94f668a86a2b9352dc22eae/core/dbt/cli/flags.py#L58) if it is not specified. We can also see that in that instantiation, `silence` is never passed 🙃 and that is indeed the issue. Interestingly, I'm not sure why the related are passing ([TestWarnErrorOptionsFromCLI.test_can_silence](https://github.com/dbt-labs/dbt-core/blob/ca163c3d6ecaf989e94f668a86a2b9352dc22eae/tests/functional/configs/test_warn_error_options.py#L51), [TestWarnErrorOptionsFromProject.test_can_silence](https://github.com/dbt-labs/dbt-core/blob/ca163c3d6ecaf989e94f668a86a2b9352dc22eae/tests/functional/configs/test_warn_error_options.py#L128))

**Comment by QMalcolm** [comment]
Back when I did the work for https://github.com/dbt-labs/dbt-core/pull/10058 (specifically https://github.com/dbt-labs/dbt-core/commit/c52d6531fcc0ca51804906dccb726d68be8550ea) I thought that the `warn_error_options` would automatically be converted from the yaml to the `WarnErrorOptions` object as we were building the `ProjectFlags` object, which holds `warn_error_options`, via `ProjectFlags.from_dict`. And I thought this was validated by the `test_can_silence` test added in c52d6531fcc0ca51804906dccb726d68be8550ea. However, there were two problems:

1. The definition of `warn_error_options` on `PrjectFlags` is a dict, not a `WarnErrorOptions` object
2. The `test_can_silence` test was broken, and not testing what I thought

The quick fix (#10359) is to ensure `silence` is passed to `WarnErrorOptions` instantiation in `dbt.cli.flags.convert_config`. The better fix would be to make the `warn_error_options` of `ProjectFlags` a `WarnErrorOptions` object instead of a dict. However, to do this we first need to update dbt-common's `WarnErrorOptions` definition to default `include` to an empty list. Doing so would allow us to get rid of `convert_config` entirely.
